File: iosApp/iosApp/notes/presentation/components/NotesListComponents.swift
import SwiftUI

struct NotesHeader: View {
    let titleColor: Color
    let metaColor: Color
    let summaryText: String
    let statusLabel: String

    var body: some View {
        VStack(alignment: .leading, spacing: 10) {
            Text("Notes")
                .font(.voraHeadlineLarge)
                .foregroundStyle(titleColor)

            HStack(spacing: 10) {
                Text(summaryText)
                    .font(.voraLabel)
                    .foregroundStyle(metaColor)
                StatusChip(
                    label: statusLabel,
                    background: statusLabel == "Syncing" ? Color.voraSurfaceMuted : .voraSuccess,
                    text: statusLabel == "Syncing" ? .voraTertiary : .voraWhite,
                    iconName: statusLabel == "Syncing" ? "arrow.triangle.2.circlepath" : "checkmark.circle.fill"
                )
            }
        }
    }
}

struct NotesSearchBar: View {
    @Binding var query: String
    let textColor: Color
    let background: Color
    let border: Color

    var body: some View {
        HStack(spacing: 10) {
            Image(systemName: "magnifyingglass")
                .font(.system(size: 14, weight: .semibold))
                .foregroundStyle(textColor.opacity(0.65))

            TextField("Search transcripts", text: $query)
                .font(.voraBody)
                .foregroundStyle(textColor.opacity(0.85))
                .autocorrectionDisabled()
                .textInputAutocapitalization(.never)

            if !query.isEmpty {
                Button(action: { query = "" }) {
                    Image(systemName: "xmark.circle.fill")
                        .foregroundStyle(textColor.opacity(0.65))
                }
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(.horizontal, VoraSpacing.searchHorizontal)
        .padding(.vertical, VoraSpacing.searchVertical)
        .background(background)
        .overlay(
            RoundedRectangle(cornerRadius: 16)
                .stroke(border, lineWidth: 1)
        )
        .clipShape(RoundedRectangle(cornerRadius: 16))
    }
}

struct NotesFilterRow: View {
    let filters: [NotesSourceFilterItem]
    let onSelect: (String) -> Void

    var body: some View {
        HStack(spacing: 8) {
            ForEach(filters) { filter in
                FilterChip(
                    label: "\(filter.label) \(filter.count)",
                    selected: filter.selected,
                    iconName: filter.iconName
                )
                .onTapGesture {
                    onSelect(filter.key)
                }
            }
        }
    }
}

struct NotesLoadingFilterRow: View {
    let background: Color
    let selectedBackground: Color

    var body: some View {
        HStack(spacing: 8) {
            LoadingPill(width: 48, background: selectedBackground)
            LoadingPill(width: 70, background: background)
            LoadingPill(width: 82, background: Color(hex: 0xF8ECE9))
        }
    }
}

struct NotesRecentLabel: View {
    let textColor: Color

    var body: some View {
        Text("RECENT")
            .font(.voraLabelSmall)
            .foregroundStyle(textColor)
    }
}

struct FilterChip: View {
    let label: String
    let selected: Bool
    var background: Color = Color(hex: 0xEFF4F7)
    var text: Color = .voraLogoInk
    var iconName: String? = nil

    var body: some View {
        HStack(spacing: 6) {
            if let iconName {
                Image(systemName: iconName)
                    .font(.system(size: 11, weight: .semibold))
            }
            Text(label)
                .font(.voraLabel)
        }
        .foregroundStyle(selected ? .voraWhite : text)
        .padding(.horizontal, VoraSpacing.chipHorizontal)
        .padding(.vertical, VoraSpacing.chipVertical)
        .background(selected ? Color.voraLogoInk : background)
        .clipShape(Capsule())
    }
}

struct StatusChip: View {
    let label: String
    let background: Color
    let text: Color
    var iconName: String? = nil

    var body: some View {
        HStack(spacing: 6) {
            if let iconName {
                Image(systemName: iconName)
                    .font(.system(size: 11, weight: .semibold))
            }
            Text(label)
                .font(.voraLabel)
        }
        .foregroundStyle(text)
        .padding(.horizontal, VoraSpacing.chipHorizontal)
        .padding(.vertical, VoraSpacing.chipVertical)
        .background(background)
        .clipShape(Capsule())
    }
}

extension String {
    var sourceIconName: String {
        switch self {
        case "Phone":
            return "iphone"
        case "Watch":
            return "applewatch"
        case "Car":
            return "car.fill"
        default:
            return "circle.fill"
        }
    }
}

#Preview("Status Synced") {
    StatusChip(
        label: "Synced",
        background: .voraSuccess,
        text: .voraWhite,
        iconName: "checkmark.circle.fill"
    )
    .padding()
    .background(Color.voraPaper)
}

#Preview("Status Pending") {
    FilterChip(
        label: "Pending 2",
        selected: false,
        background: Color(hex: 0xEFF4F7),
        text: .voraLogoInk,
        iconName: "arrow.triangle.2.circlepath"
    )
    .padding()
    .background(Color.voraPaper)
}

#Preview("Status Needs Review") {
    FilterChip(
        label: "Needs review",
        selected: false,
        background: Color(hex: 0xF9ECE8),
        text: Color(hex: 0xC95B4A),
        iconName: "exclamationmark.circle"
    )
    .padding()
    .background(Color.voraPaper)
}

#Preview("Status Source Phone") {
    StatusChip(
        label: "Phone",
        background: Color(hex: 0xEFF4F7),
        text: .voraLogoInk,
        iconName: "iphone"
    )
    .padding()
    .background(Color.voraPaper)
}

#Preview("Status Source Watch") {
    StatusChip(
        label: "Watch",
        background: Color(hex: 0xEFF4F7),
        text: .voraLogoInk,
        iconName: "applewatch"
    )
    .padding()
    .background(Color.voraPaper)
}

#Preview("Status Source Car") {
    StatusChip(
        label: "Car",
        background: Color(hex: 0xEFF4F7),
        text: .voraLogoInk,
        iconName: "car.fill"
    )
    .padding()
    .background(Color.voraPaper)
}

File: composeApp/src/androidMain/kotlin/com/mohamedfaridelsherbini/vora/notes/presentation/components/NotesListComponents.kt
package com.mohamedfaridelsherbini.vora.notes.presentation.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.CheckCircle
import androidx.compose.material.icons.outlined.Close
import androidx.compose.material.icons.outlined.DirectionsCar
import androidx.compose.material.icons.outlined.PhoneAndroid
import androidx.compose.material.icons.outlined.PriorityHigh
import androidx.compose.material.icons.outlined.Search
import androidx.compose.material.icons.outlined.Sync
import androidx.compose.material.icons.outlined.Watch
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.mohamedfaridelsherbini.vora.notes.presentation.state.NotesSourceFilterUi
import com.mohamedfaridelsherbini.vora.presentation.theme.VoraColors
import com.mohamedfaridelsherbini.vora.presentation.theme.VoraSpacing
import com.mohamedfaridelsherbini.vora.presentation.theme.interFontFamily

@Composable
internal fun NotesHeader(
    titleColor: Color,
    subtitleColor: Color,
    summaryText: String,
    statusLabel: String,
    statusBackground: Color,
    statusTextColor: Color,
) {
    androidx.compose.foundation.layout.Column(verticalArrangement = Arrangement.spacedBy(10.dp)) {
        Text(
            text = "Notes",
            color = titleColor,
            fontFamily = interFontFamily(),
            fontWeight = FontWeight.Bold,
            fontSize = 34.sp,
        )
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(10.dp),
        ) {
            Text(
                text = summaryText,
                color = subtitleColor,
                fontFamily = interFontFamily(),
                fontWeight = FontWeight.Medium,
                fontSize = 12.sp,
            )
            StatusChip(
                label = statusLabel,
                background = statusBackground,
                textColor = statusTextColor,
                icon = if (statusLabel == "Syncing") Icons.Outlined.Sync else Icons.Outlined.CheckCircle,
            )
        }
    }
}

@Composable
internal fun NotesSearchBar(
    query: String,
    onQueryChange: (String) -> Unit,
    background: Color,
    borderColor: Color,
    textColor: Color,
    iconColor: Color,
) {
    BasicTextField(
        value = query,
        onValueChange = onQueryChange,
        textStyle = androidx.compose.ui.text.TextStyle(
            color = textColor.copy(alpha = 0.85f),
            fontFamily = interFontFamily(),
            fontWeight = FontWeight.Medium,
            fontSize = 14.sp,
        ),
        singleLine = true,
        cursorBrush = SolidColor(textColor),
        decorationBox = { innerTextField ->
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .background(background, RoundedCornerShape(16.dp))
                    .border(1.dp, borderColor, RoundedCornerShape(16.dp))
                    .padding(horizontal = VoraSpacing.SearchHorizontal, vertical = VoraSpacing.SearchVertical),
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(10.dp),
            ) {
                Icon(
                    imageVector = Icons.Outlined.Search,
                    contentDescription = null,
                    tint = iconColor.copy(alpha = 0.65f),
                    modifier = Modifier.size(16.dp),
                )
                Box(modifier = Modifier.weight(1f)) {
                    if (query.isEmpty()) {
                        Text(
                            text = "Search transcripts",
                            color = iconColor.copy(alpha = 0.65f),
                            fontFamily = interFontFamily(),
                            fontWeight = FontWeight.Medium,
                            fontSize = 14.sp,
                        )
                    }
                    innerTextField()
                }
                if (query.isNotEmpty()) {
                    IconButton(
                        onClick = { onQueryChange("") },
                        modifier = Modifier.size(16.dp),
                    ) {
                        Icon(
                            imageVector = Icons.Outlined.Close,
                            contentDescription = "Clear search",
                            tint = iconColor.copy(alpha = 0.65f),
                            modifier = Modifier.size(16.dp),
                        )
                    }
                }
            }
        },
    )
}

@Composable
internal fun NotesFilterRow(
    filters: List<NotesSourceFilterUi>,
    background: Color,
    textColor: Color,
    onSelect: (String) -> Unit,
) {
    Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
        filters.forEach { filter ->
            FilterChip(
                label = "${filter.label} ${filter.count}",
                selected = filter.selected,
                background = background,
                textColor = textColor,
                icon = filter.icon(),
                onClick = { onSelect(filter.key) },
            )
        }
    }
}

@Composable
internal fun NotesLoadingFilterRow(
    background: Color,
    selectedBackground: Color,
) {
    Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
        LoadingPill(width = 48.dp, background = selectedBackground)
        LoadingPill(width = 70.dp, background = background)
        LoadingPill(width = 82.dp, background = Color(0xFFF8ECE9))
    }
}

@Composable
internal fun NotesRecentLabel(textColor: Color) {
    Text(
        text = "RECENT",
        color = textColor,
        fontFamily = interFontFamily(),
        fontWeight = FontWeight.SemiBold,
        fontSize = 11.sp,
    )
}

@Composable
private fun StatusChip(
    label: String,
    background: Color,
    textColor: Color,
    icon: ImageVector? = null,
) {
    Row(
        modifier = Modifier
            .background(background, RoundedCornerShape(999.dp))
            .padding(horizontal = VoraSpacing.ChipHorizontal, vertical = VoraSpacing.ChipVertical),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(6.dp),
    ) {
        if (icon != null) {
            Icon(
                imageVector = icon,
                contentDescription = null,
                tint = textColor,
                modifier = Modifier.size(12.dp),
            )
        }
        Text(
            text = label,
            color = textColor,
            fontFamily = interFontFamily(),
            fontWeight = FontWeight.SemiBold,
            fontSize = 12.sp,
        )
    }
}

@Composable
private fun FilterChip(
    label: String,
    selected: Boolean,
    background: Color,
    textColor: Color,
    icon: ImageVector? = null,
    onClick: (() -> Unit)? = null,
) {
    Row(
        modifier = Modifier
            .background(
                color = if (selected) VoraColors.LogoInk else background,
                shape = RoundedCornerShape(999.dp),
            )
            .then(if (onClick != null) Modifier.clickable(onClick = onClick) else Modifier)
            .padding(horizontal = VoraSpacing.ChipHorizontal, vertical = VoraSpacing.ChipVertical),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        val foreground = if (selected) VoraColors.LogoPaper else textColor
        if (icon != null) {
            Icon(
                imageVector = icon,
                contentDescription = null,
                tint = foreground,
                modifier = Modifier.size(12.dp),
            )
            Spacer(modifier = Modifier.size(6.dp))
        }
        Text(
            text = label,
            color = foreground,
            fontFamily = interFontFamily(),
            fontWeight = FontWeight.SemiBold,
            fontSize = 12.sp,
        )
    }
}

@Composable
private fun LoadingPill(
    width: androidx.compose.ui.unit.Dp,
    background: Color,
) {
    Box(
        modifier = Modifier
            .size(width = width, height = 22.dp)
            .background(background, RoundedCornerShape(999.dp)),
    )
}

private fun NotesSourceFilterUi.icon(): ImageVector = when (key) {
    "phone" -> Icons.Outlined.PhoneAndroid
    "smart" -> Icons.Outlined.Watch
    "car" -> Icons.Outlined.DirectionsCar
    else -> Icons.Outlined.CheckCircle
}

@Preview(name = "Status Synced", showBackground = true, backgroundColor = 0xFFF9FAFB)
@Composable
private fun SyncedStatusChipPreview() {
    StatusChip(
        label = "Synced",
        background = Color(0xFF3A7D58),
        textColor = VoraColors.VoraWhite,
        icon = Icons.Outlined.CheckCircle,
    )
}

@Preview(name = "Status Pending", showBackground = true, backgroundColor = 0xFFF9FAFB)
@Composable
private fun PendingStatusChipPreview() {
    FilterChip(
        label = "Pending 2",
        selected = false,
        background = Color(0xFFEFF4F7),
        textColor = VoraColors.LogoInk,
        icon = Icons.Outlined.Sync,
        onClick = {},
    )
}

@Preview(name = "Status Needs Review", showBackground = true, backgroundColor = 0xFFF9FAFB)
@Composable
private fun NeedsReviewStatusChipPreview() {
    FilterChip(
        label = "Needs review",
        selected = false,
        background = Color(0xFFF9ECE8),
        textColor = Color(0xFFC95B4A),
        icon = Icons.Outlined.PriorityHigh,
        onClick = {},
    )
}

@Preview(name = "Status Source Phone", showBackground = true, backgroundColor = 0xFFF9FAFB)
@Composable
private fun SourcePhoneChipPreview() {
    StatusChip(
        label = "Phone",
        background = Color(0xFFEFF4F7),
        textColor = VoraColors.LogoInk,
        icon = Icons.Outlined.PhoneAndroid,
    )
}

@Preview(name = "Status Source Watch", showBackground = true, backgroundColor = 0xFFF9FAFB)
@Composable
private fun SourceWatchChipPreview() {
    StatusChip(
        label = "Watch",
        background = Color(0xFFEFF4F7),
        textColor = VoraColors.LogoInk,
        icon = Icons.Outlined.Watch,
    )
}

@Preview(name = "Status Source Car", showBackground = true, backgroundColor = 0xFFF9FAFB)
@Composable
private fun SourceCarChipPreview() {
    StatusChip(
        label = "Car",
        background = Color(0xFFEFF4F7),
        textColor = VoraColors.LogoInk,
        icon = Icons.Outlined.DirectionsCar,
    )
}
